pandas/core/arrays/_reshaping.py

```diff
@@ -3,43 +3,97 @@
 """
 from functools import wraps
 from typing import Tuple
+import warnings
 
 import numpy as np
 
 from pandas._libs.lib import is_integer
 
+msg = (
+    "ExtensionArray subclass {name} defines {method}. "
+    "In the future, this will not be supported. Please "
+    "inherit {method} from the ExtensionArray base class."
+)
+
 
 def implement_2d(cls):
     """
-    A decorator to take a 1-dimension-only ExtensionArray subclass and make
+    Patch a 1-dimension-only ExtensionArray subclass and make
     it support limited 2-dimensional operations.
-    """
-    from pandas.core.arrays import ExtensionArray
 
-    # For backwards-compatibility, if an EA author implemented __len__
-    #  but not size, we use that __len__ method to get an array's size.
+    We achieve this by rewriting dimension-dependent methods to
+    pre-process the inputs to make them look 1d, call the underlying
+    method, and post-process the output.
+    """
+    if cls._allows_2d:
+        return
+
+    if cls.__name__ == "ExtensionArray" and cls.__module__ == "pandas.core.arrays.base":
+        # No need to patch for ExtensionArray base class.
+        return
+    else:
+        from pandas.core.arrays import ExtensionArray
+
+    # For backwards-compatibility, we use the length, size, or shape
+    # defined by the subclass. We can always define the other two in
+    # terms of the one.
     has_size = cls.size is not ExtensionArray.size
     has_shape = cls.shape is not ExtensionArray.shape
-    has_len = cls.__len__ is not ExtensionArray.__len__
 
-    if not has_size and has_len:
-        cls.size = property(cls.__len__)
-        cls.__len__ = ExtensionArray.__len__
+    orig_len = cls.__len__
+    # TODO: Find a better way to do this. I suspect we could check whether
+    # our cls.bases contains ExtensionArray...
+    if hasattr(orig_len, "_original_len"):
+        # When a user does class Foo(Bar(ExtensionArray)):
+        # we want to use the unpatched verison.
+        orig_len = orig_len._original_len
+
+    orig_shape = cls.shape
+
+    @wraps(orig_len)
+    def __len__(self):
+        length = orig_len(self)
+        if self._ExtensionArray__expanded_dim is None:
+            result = length
+        elif self._ExtensionArray__expanded_dim == 0:
+            result = length
+        else:
+            result = 1
```

Comment:
im not sure i follow.  is expanded_dim an indicator or is it a patched ndim or something else?  Is 1 hard-coded here because we support only (N, 1) and (1, N)?

Reply:
`expanded_dim` is an indicator.

If `array._expanded_dim` == 1, that means  `array.shape` is `(1, N)` and `len(array)` is `1`.

Reply:
(i see this is clarified below)